packages/cli/src/upgrade/upgrade.js

```diff
@@ -4,151 +4,215 @@
-        "'package.json' file. Aborting."
+  if (!newVersion) {
+    logger.error(
+      `Provided version "${newVersion}" is not allowed. Please pass a valid semver version`
     );
-    return;
+    return null;
   }
 
-  logger.info(
-    `Upgrading project to react-native v${installed.version}\n` +
-      `Check out the release notes and breaking changes: ` +
-      `https://github.com/facebook/react-native/releases/tag/v${semver.major(
-        v
-      )}.${semver.minor(v)}.0`
-  );
-
-  // >= v0.21.0, we require react to be a peer dependency
-  if (semver.gte(v, '0.21.0') && !packageJSON.dependencies.react) {
-    warn(
-      'Your "package.json" file doesn\'t seem to have "react" as a dependency.\n' +
-        '"react" was changed from a dependency to a peer dependency in react-native v0.21.0.\n' +
-        'Therefore, it\'s necessary to include "react" in your project\'s dependencies.\n' +
-        'Please run "npm install --save react", then re-run "react-native upgrade".\n'
+  if (currentVersion > newVersion) {
+    logger.error(
+      `Trying to upgrade from newer version "${currentVersion}" to older "${newVersion}"`
     );
-    return;
+    return null;
   }
 
-  if (semver.satisfies(v, '~0.26.0')) {
-    warn(
-      'React Native 0.26 introduced some breaking changes to the native files on iOS. You can\n' +
-        'perform them manually by checking the release notes or use "rnpm" ' +
-        'to do it automatically.\n' +
-        'Just run:\n' +
-        '"npm install -g rnpm && npm install rnpm-plugin-upgrade@0.26 --save-dev", ' +
-        'then run "rnpm upgrade".'
+  if (currentVersion === newVersion) {
+    const {
+      dependencies: { 'react-native': version },
+    } = require(path.join(projectDir, 'package.json'));
+    if (semver.satisfies(newVersion, version)) {
+      logger.warn(
+        `Specified version "${newVersion}" is already installed in node_modules and it satisfies "${version}" semver range. No need to upgrade`
+      );
+      return null;
+    }
+    logger.error(
+      `Dependency mismatch. Specified version "${newVersion}" is already installed in node_modules and it doesn't satisfy "${version}" semver range of your "react-native" dependency. Please re-install your dependencies`
     );
+    return null;
   }
 
-  upgradeProjectFiles(projectDir, projectName);
+  return newVersion;
+};
 
+const installDeps = async (newVersion, projectDir) => {
   logger.info(
-    `Successfully upgraded this project to react-native v${installed.version}`
+    `Installing react-native@${newVersion} and its peer dependencies...`
   );
-}
+  const peerDeps = await getRNPeerDeps(newVersion);
+  const pm = new PackageManager({ projectDir });
+  const deps = [
+    `react-native@${newVersion}`,
+    ...Object.entries(peerDeps).map(
+      // $FlowFixMe - Object.entries type definition is poor
+      ([module, version]) => `${module}@${version}`
+    ),
```

Comment:
I recommend using Object.keys and then using the key to index back in. That’s how we work around this internally.